### PR TITLE
Use typed-rest-client 1.0.7 to allow toolLib.downloadTool to work with redirects.

### PR DESCRIPTION
**Pull request by sanderaernouts**
# Background
While trying to download a tools exe from Github releases using toolLib.downloadTool I got the following error: UnhandledPromiseRejectionWarning: Error: Unexpected HTTP response: 302. Github was redirecting met to an AWS blob which seems perfectly fine to me but toolLib.downloadTool only accepts HTTP 200.

# Change
I updated typed-rest-client to 1.0.7 to allow toolLib.downloadTool to work with HTTP redirects.

### CODE REVIEW

**Comment by stephenmichaelf** [comment]
@sanderaernouts Thanks for taking the time to do a PR, we appreciate it! This module relies on the typed-rest-client which we also own. We recently made changes [here](https://github.com/Microsoft/typed-rest-client/blob/dffabe6a69a456c1672dae7e1b86d7a37f9017f6/lib/HttpClient.ts#L212) to have that client handle redirects. It's a lower level client that should have that responsibility. I think if we simply update this module to point to the latest typed-rest-client it should resolve your issue.

Since you already started working on this and investigating do you want to give that a try and see if it works?

**Comment by sanderaernouts** [comment]
@stephenmichaelf  sure thing. I updated typed-rest-client to 1.0.7 and added some test using https://httpbin.org to test the redirect status codes work with downloadTool.

**Comment by sanderaernouts** [comment]
@stephenmichaelf sure thing. I updated typed-rest-client to 1.0.7 and added some tests to verify the redirect status codes work with downloadTool. I did not change the statuscode back to statuscode == 200 because there are other 2xx status codes that indicate success.

File: tool.ts [hunk 1]
```diff
@@ -222,7 +222,7 @@ export async function downloadTool(url: string, fileName?: string): Promise<stri
             // TODO: retries
             tl.debug('downloading');
             let response: httpm.HttpClientResponse = await http.get(url);
-            if (response.message.statusCode != 200) {
+            if (response.message.statusCode >= 400) {
```
Comment (bryanmacfarlane) [inline tool.ts]: We don't need this if typed rest client is handling 3xx.  if there's an large / infinite redirect issue, typed rest client will appropriately error and this shouldn't mask it.  Same for 500s.

File: package.json [hunk 2]
```diff
@@ -25,7 +25,7 @@
     "semver": "^5.3.0",
     "@types/semver": "^5.3.0",
     "semver-compare": "^1.0.0",
-    "typed-rest-client": "0.14.1",
+    "typed-rest-client": "1.0.7",
```
Comment (bryanmacfarlane) [inline package.json]: This is good.

File: .gitignore [hunk 3]
```diff
@@ -34,8 +34,17 @@ build/Release
 node_modules
 jspm_packages
 
+# Npm lock files
+package-lock.json
```
Comment (bryanmacfarlane) [inline .gitignore]: this gets checked in and it shouldn't be ignored.  If this changes you should see the diff

File: test/toolTests.ts [hunk 4]
```diff
@@ -81,24 +81,83 @@ describe('Tool Tests', function () {
...
 
-    it('has status code in exception dictionary for non 200 response', function () {
-        this.timeout(2000);
 
-        return new Promise<void>(async(resolve, reject)=> {
-            try {
-                let nonExistentUrl: string = "http://httpbin.org/nonexistent";
-                let downPath: string = await toolLib.downloadTool(nonExistentUrl);
+    
+    describe('has status code in exception dictionary for HTTP error code responses', function () {
```
Comment (bryanmacfarlane) [inline test/toolTests.ts]: I think the only test we need is a download that redirects (once) and we make sure it's 200.  The tests below are the responsibility of the typed rest client

File: .gitignore [hunk 5]
```diff
@@ -34,8 +34,17 @@ build/Release
...
 .npm
 
 # Optional REPL history
 .node_repl_history
+
+# Temp directories
+TEMP
+
+# Resources
+Strings/*
```
Comment (stephenmichaelf) [inline .gitignore]: We need these files.

File: test/toolTests.ts [hunk 6]
```diff
@@ -81,24 +81,83 @@ describe('Tool Tests', function () {
...
-            catch (err){
-                assert.equal(err['httpStatusCode'], 404, 'status code exists');
 
-                resolve();
-            }
-        });
+        
     });
 
+    describe('accepts any status code < 400', function() {
```
Comment (stephenmichaelf) [inline test/toolTests.ts]: Can remove this test.

Comment (stephenmichaelf) [reply test/toolTests.ts]: Or modify to what Bryan said, check that for a redirect URL we get the final URL and a 200.

File: .gitignore (hunk 5 above)
Comment (sanderaernouts) [reply .gitignore]: Thought I removed that one >..<

Comment (sanderaernouts) [reply .gitignore]: Does TEMP need to be removed from the .gitignore as well?

**Comment by stephenmichaelf** [comment]
Thanks for taking the time to make these adjustments, it is much appreciated.

LGTM. 

@bryanmacfarlane if you agree I am fine merging this.

**Comment by stephenmichaelf** [comment]
Bump @bryanmacfarlane.

**Comment by bryanmacfarlane** [comment]
Thanks!

**Comment by bryanmacfarlane** [comment]
I will release after lunch :)

**Comment by bryanmacfarlane** [comment]
I published 0.5.1

**Comment by stephenmichaelf** [comment]
Thanks!

**Pull request merged**